src/core/lib/event_engine/ares_driver.cc

```diff
@@ -0,0 +1,952 @@
+        FdNode* ret = node->next;
+        node->next = node->next->next;
+        head_ = phony_head.next;
+        return ret;
+      }
+      node = node->next;
+    }
+    return nullptr;
+  }
+
+ private:
+  FdNode* head_ = nullptr;
+};
+
+GrpcAresRequest::GrpcAresRequest(
+    absl::string_view name, absl::optional<absl::string_view> default_port,
+    EventEngine::Duration timeout,
+    RegisterAresSocketWithPollerCallback register_cb, EventEngine* event_engine)
+    : grpc_core::RefCounted<GrpcAresRequest>(
+          GRPC_TRACE_FLAG_ENABLED(grpc_trace_ares_driver) ? "GrpcAresRequest"
+                                                          : nullptr),
+      name_(name),
+      default_port_(default_port.has_value() ? *default_port : ""),
+      timeout_(timeout),
+      fd_node_list_(std::make_unique<FdNodeList>()),
+      event_engine_(event_engine),
+      polled_fd_factory_(NewGrpcPolledFdFactory(std::move(register_cb), &mu_)) {
+}
+
+GrpcAresRequest::~GrpcAresRequest() {
+  if (initialized_) {
+    ares_destroy(channel_);
+    GRPC_ARES_DRIVER_TRACE_LOG("request:%p destructor", this);
+    GRPC_ARES_DRIVER_STACK_TRACE();
+  }
+}
+
+absl::Status GrpcAresRequest::Initialize(absl::string_view dns_server,
+                                         bool check_port) {
+  absl::MutexLock lock(&mu_);
+  GPR_DEBUG_ASSERT(!initialized_);
+  absl::string_view port;
+  GPR_ASSERT(grpc_core::SplitHostPort(name_, &host_, &port));
+  absl::Status error;
+  if (host_.empty()) {
+    error =
+        grpc_error_set_str(GRPC_ERROR_CREATE("unparseable host:port"),
+                           grpc_core::StatusStrProperty::kTargetAddress, name_);
+    return error;
+  } else if (check_port && port.empty()) {
+    if (default_port_.empty()) {
+      error = grpc_error_set_str(GRPC_ERROR_CREATE("no port in name"),
+                                 grpc_core::StatusStrProperty::kTargetAddress,
+                                 name_);
+      return error;
+    }
+    port = default_port_;
+  }
+  if (!port.empty()) {
+    port_ = grpc_strhtons(std::string(port).c_str());
+  }
+  ares_options opts = {};
+  opts.flags |= ARES_FLAG_STAYOPEN;
+  int status = ares_init_options(&channel_, &opts, ARES_OPT_FLAGS);
+  if (status != ARES_SUCCESS) {
+    gpr_log(GPR_ERROR, "ares_init_options failed, status: %d", status);
+    return GRPC_ERROR_CREATE(absl::StrCat(
+        "Failed to init ares channel. C-ares error: ", ares_strerror(status)));
+  }
+  ares_driver_test_only_inject_config(channel_);
+  error = SetRequestDNSServer(dns_server);
+  if (!error.ok()) {
+    return error;
+  }
+  initialized_ = true;
+  return absl::OkStatus();
+}
+
+void GrpcAresRequest::Cancel() {
+  absl::MutexLock lock(&mu_);
+  if (!shutting_down_) {
+    shutting_down_ = true;
+    cancelled_ = true;
+    ShutdownPollerHandles();
+  }
+}
+
+void GrpcAresRequest::Work() {
+  std::unique_ptr<FdNodeList> new_list = std::make_unique<FdNodeList>();
```

Comment:
nit: `auto new_list = std::make_unique<FdNodeList>();` is sufficiently self-descriptive for `std::make_*` calls.